File: clients/dotnet/src/Ioka.Munarium.Client/SseParser.cs
// SPDX-License-Identifier: Apache-2.0
// A minimal incremental Server-Sent-Events parser for the streaming turn
// plane (POST /v1/sessions/{id}/turns/stream). Hand-rolled by design — the
// server's emitter is simple (named events, single-line JSON data, comment
// keep-alives) and a dependency would be heavier than the format.
//
// The parser is PURE: feed it byte chunks as they arrive (chunk boundaries
// may fall anywhere, including mid-line and mid-UTF-8-codepoint) and it
// yields complete events. Per the SSE grammar it handles `event:`/`data:`
// fields, multi-line data accumulation, \n / \r\n / \r line endings, comment
// lines (leading ':', the keep-alive form), and ignores fields it does not
// know (`id:`, `retry:`).
//
// Retention is bounded EXACTLY: a stream that never terminates its lines or
// events cannot grow client memory past MaxEventBytes — the bound is
// checked before a byte is buffered. Overflow never discards events the
// same chunk completed: Push returns them and the parser is poisoned so the
// NEXT push throws SseOverflowException (it throws immediately only when
// nothing completed), and the caller ends the stream with a typed error
// instead of buffering toward an OOM kill.

using System.Buffers;
using System.Text;

namespace Ioka.Munarium.Client;

/// <summary>One dispatched SSE event: the event name (empty = the spec's
/// default "message") and the joined data payload.</summary>
internal readonly record struct SseEvent(string Event, string Data);

/// <summary>The parser refused to buffer further — the peer sent more than
/// <see cref="SseParser.MaxEventBytes"/> without completing an event.</summary>
internal sealed class SseOverflowException()
    : Exception("SSE event exceeded the buffer cap without completing");

internal sealed class SseParser
{
    /// <summary>Upper bound on one event's buffered bytes (pending line +
    /// accumulated data). A terminal done event carries a whole TurnResult —
    /// hits text included — so the cap is generous; anything past it is a
    /// misbehaving peer, not a real event.</summary>
    internal const int MaxEventBytes = 16 * 1024 * 1024;

    private static readonly SseEvent[] None = [];

    /// <summary>Undelivered raw bytes (a partial line, possibly a partial
    /// codepoint). Reset — not reallocated — per line, so capacity is
    /// reused; whole runs are appended, never byte-by-byte.</summary>
    private readonly ArrayBufferWriter<byte> _buf = new(256);

    /// <summary>Accumulated data: lines for the event being built.</summary>
    private readonly List<string> _data = [];

    /// <summary>Bytes across _data (tracked so the cap is O(1) to enforce).</summary>
    private int _dataBytes;

    /// <summary>The pending event: name.</summary>
    private string _event = "";

    /// <summary>True when the previous byte was CR — a following LF is the
    /// same line ending, not an extra blank line.</summary>
    private bool _sawCr;

    /// <summary>Set once the cap was hit; every later push throws.</summary>
    private bool _poisoned;

    /// <summary>Feed one chunk; returns every event completed by it, in
    /// order (a shared empty list when none — do not mutate). Throws
    /// <see cref="SseOverflowException"/> past the cap, after first
    /// returning anything the chunk completed.</summary>
    public IReadOnlyList<SseEvent> Push(ReadOnlySpan<byte> chunk)
    {
        if (_poisoned) throw new SseOverflowException();
        List<SseEvent>? output = null;
        while (!chunk.IsEmpty)
        {
            if (_sawCr && chunk[0] == (byte)'\n')
            {
                _sawCr = false; // CRLF: LF consumed
                chunk = chunk[1..];
                continue;
            }
            var end = chunk.IndexOfAny((byte)'\r', (byte)'\n');
            var run = end < 0 ? chunk : chunk[..end];
            if (_buf.WrittenCount + run.Length + _dataBytes > MaxEventBytes)
            {
                _poisoned = true;
                break;
            }
            if (end < 0)
            {
                _sawCr = false;
                _buf.Write(run);
                break;
            }
            _sawCr = chunk[end] == (byte)'\r';
            Line(run, ref output);
            _buf.ResetWrittenCount();
            chunk = chunk[(end + 1)..];
        }
        if (_poisoned && output is null) throw new SseOverflowException();
        return output ?? (IReadOnlyList<SseEvent>)None;
    }

    /// <summary>Handle one complete line: the pending buffer plus
    /// <paramref name="tail"/> (the part of it that arrived in this chunk).</summary>
    private void Line(ReadOnlySpan<byte> tail, ref List<SseEvent>? output)
    {
        ReadOnlySpan<byte> line;
        if (_buf.WrittenCount == 0)
        {
            line = tail; // whole line in this chunk: no copy
        }
        else
        {
            _buf.Write(tail);
            line = _buf.WrittenSpan;
        }
        if (line.IsEmpty)
        {
            // Blank line = dispatch. An event with no data lines is dropped
            // per the SSE spec (this is what makes comment keep-alives free).
            if (_data.Count > 0)
            {
                (output ??= []).Add(new SseEvent(_event, string.Join('\n', _data)));
                _data.Clear();
            }
            _event = "";
            _dataBytes = 0;
            return;
        }
        if (line[0] == (byte)':') return; // comment / keep-alive

        var colon = line.IndexOf((byte)':');
        var field = colon < 0 ? line : line[..colon];
        var value = colon < 0 ? default : line[(colon + 1)..];
        if (!value.IsEmpty && value[0] == (byte)' ') value = value[1..];
        // Only the value is decoded; the field name is matched as bytes.
        if (field.SequenceEqual("event"u8))
        {
            _event = Encoding.UTF8.GetString(value);
        }
        else if (field.SequenceEqual("data"u8))
        {
            _dataBytes += value.Length;
            _data.Add(Encoding.UTF8.GetString(value));
        }
        // id / retry / unknown fields: ignored
    }
}

File: clients/java/src/main/java/io/ioka/munarium/client/rest/SseParser.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.munarium.client.rest;

import java.io.ByteArrayOutputStream;
import java.nio.charset.StandardCharsets;
import java.util.ArrayList;
import java.util.List;

/**
 * A minimal incremental Server-Sent-Events parser for the streaming turn
 * plane. Hand-rolled by design — the server's emitter is simple (named
 * events, single-line JSON data, comment keep-alives) and a dependency
 * would be heavier than the format.
 *
 * <p>PURE: feed byte chunks as they arrive (boundaries may fall anywhere,
 * including mid-line and mid-UTF-8-codepoint) and it yields complete
 * events. Handles {@code event:}/{@code data:} fields, multi-line data
 * accumulation, {@code \n}/{@code \r\n}/{@code \r} line endings, comment
 * keep-alives, and unknown fields ({@code id:}, {@code retry:}) ignored.
 *
 * <p>Retention is bounded: past {@link #MAX_EVENT_BYTES} of buffered bytes
 * without a completed event, {@link Overflow} is thrown — a misbehaving
 * peer must not grow client memory toward an OOM kill.
 */
final class SseParser {
    /** Generous — a terminal done event carries a whole TurnResponse. */
    static final int MAX_EVENT_BYTES = 16 * 1024 * 1024;

    record Event(String event, String data) {}

    static final class Overflow extends RuntimeException {
        Overflow() {
            super("SSE peer exceeded the event buffer without completing an event");
        }
    }

    private final ByteArrayOutputStream buf = new ByteArrayOutputStream();
    private final List<String> data = new ArrayList<>();
    private int dataBytes;
    private String event = "";
    private boolean sawCr;
    /** The cap was exceeded on an earlier push: events completed in that
     * push were still delivered (a done beside oversized trailing bytes is a
     * real result), and the NEXT push reports the overflow. */
    private boolean poisoned;

    /** Feed one chunk; returns every event completed by it, in order. */
    List<Event> push(byte[] chunk, int len) {
        if (poisoned) {
            throw new Overflow();
        }
        List<Event> out = new ArrayList<>();
        for (int i = 0; i < len; i++) {
            byte b = chunk[i];
            if (b == '\n' && sawCr) {
                sawCr = false; // CRLF: LF consumed
            } else if (b == '\r' || b == '\n') {
                sawCr = b == '\r';
                line(buf.toByteArray(), out);
                buf.reset();
            } else {
                sawCr = false;
                buf.write(b);
            }
        }
        if (buf.size() + dataBytes > MAX_EVENT_BYTES) {
            poisoned = true;
            if (out.isEmpty()) {
                throw new Overflow();
            }
        }
        return out;
    }

    private void line(byte[] raw, List<Event> out) {
        if (raw.length == 0) {
            // Blank line = dispatch. An event with no data lines is dropped
            // per the SSE spec (comment keep-alives are free); the stale
            // event name must not leak into the next event.
            if (!data.isEmpty()) {
                out.add(new Event(event, String.join("\n", data)));
                data.clear();
            }
            event = "";
            dataBytes = 0;
            return;
        }
        if (raw[0] == ':') {
            return; // comment / keep-alive
        }
        // The retention cap counts BYTES (raw line length — a slight
        // overcount including the field prefix, fine for a bound): counting
        // String.length() chars would let multi-byte payloads exceed the
        // documented cap ~3-4x before tripping.
        String text = new String(raw, StandardCharsets.UTF_8);
        int colon = text.indexOf(':');
        String field = colon < 0 ? text : text.substring(0, colon);
        String value = colon < 0 ? "" : text.substring(colon + 1);
        if (value.startsWith(" ")) {
            value = value.substring(1);
        }
        switch (field) {
            case "event" -> event = value;
            case "data" -> {
                dataBytes += raw.length;
                data.add(value);
            }
            default -> {
                // id / retry / unknown fields: ignored
            }
        }
    }
}
